qiskit-qir/src/qiskit_qir/elements.py

```diff
@@ -0,0 +1,102 @@
+##
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+##
+from dis import Instruction
+from typing import List, Union
+from cirq import num_qubits
+from qiskit import ClassicalRegister, QuantumRegister
+from qiskit.circuit.bit import Bit
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+
+
+class _QuantumCircuitElement:    
+    @classmethod
+    def from_element_list(cls, elements):
+        return [cls(elem) for elem in elements]
+
+
+class _Register(_QuantumCircuitElement):    
+    def __init__(self, register: Union[QuantumRegister, ClassicalRegister]):
+        self._register = register
+
+    def accept(self, visitor):
+        visitor.visit_register(self._register)
+
+
+class _Instruction(_QuantumCircuitElement):
+    def __init__(
+        self,
+        instruction: Instruction,
+        qargs: List[Bit],
+        cargs: List[Bit]
+    ):
+        self._instruction = instruction
+        self._qargs = qargs
+        self._cargs = cargs
+
+    def accept(self, visitor):
+        visitor.visit_instruction(self._instruction, self._qargs, self._cargs)
+
+
+class QiskitModule:
+    def __init__(self, name, num_qubits, num_clbits, elements):
+        self._name = name
+        self._elements = elements
+        self._num_qubits = num_qubits
+        self._num_clbits = num_clbits
+    
+    @property
+    def name(self):
+        return self._name
+    
+    @property
+    def num_qubits(self):
+        return self._num_qubits
+    
+    @property
+    def num_clbits(self):
+        return self._num_clbits
+
+    @classmethod
+    def from_quantum_circuit(cls, circuit: QuantumCircuit) -> "QuantumCircuit":
```

Comment:
You may be able to use `from __future__ import annotations` to enable deferred execution of type hints without using the older trick of using strings as types. That future feature was added in Python 3.7, so I think you're good to take a hard dependency on it, given your `python_requires` above.